### ISSUE TITLE
Cannot save empty value in domain settings tab

### ISSUE THREAD

**Issue by vladimirpetrov** [apache/cloudstack#5545]
##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
UI, API
~~~

##### CLOUDSTACK VERSION

~~~
4.16
~~~

##### CONFIGURATION

N/A

##### OS / ENVIRONMENT

N/A

##### SUMMARY

There is no way to delete a domain setting once there is something set.

##### STEPS TO REPRODUCE

~~~
1. Go to Domains > ROOT (for example) and click on the 'Settings' tab.
2. Set ldap.bind.password (or other setting) to some value, for example 'password'
3. Now edit the same field again, clearing the text and click on the 'OK' icon

or use the cloudmonkey command:
update configuration domainid=<domain_uuid> name=ldap.bind.password value=password
update configuration domainid=<domain_uuid> name=ldap.bind.password value=
~~~

##### EXPECTED RESULTS

~~~
The value must be cleared.
~~~

##### ACTUAL RESULTS

~~~
When using cloudmonkey, it gives the following error:

🙈 Error: (HTTP 530, error code 4250) DB Exception on: com.mysql.cj.jdbc.ClientPreparedStatement: UPDATE domain_details SET domain_details.value=null WHERE domain_details.id = 1 

This is the corresponding SQL exception in the management log:

2021-10-01 14:19:53,439 ERROR [c.c.a.ApiServer] (qtp220309324-18:ctx-b48f70c3 ctx-ec3c314b) (logid:98017a9d) unhandled exception executing api command: [Ljava.lang.String;@489189f7
com.cloud.utils.exception.CloudRuntimeException: DB Exception on: com.mysql.cj.jdbc.ClientPreparedStatement: UPDATE domain_details SET domain_details.value=null WHERE domain_details.id = 1 
        at com.cloud.utils.db.GenericDaoBase.update(GenericDaoBase.java:846)
        at com.cloud.utils.db.GenericDaoBase.update(GenericDaoBase.java:801)
        at com.cloud.utils.db.GenericDaoBase.update(GenericDaoBase.java:1361)
        at jdk.internal.reflect.GeneratedMethodAccessor103.invoke(Unknown Source)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:344)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:198)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
        at com.cloud.utils.db.TransactionContextInterceptor.invoke(TransactionContextInterceptor.java:34)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:175)
        at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:97)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
        at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:215)
        at com.sun.proxy.$Proxy165.update(Unknown Source)
        at com.cloud.configuration.ConfigurationManagerImpl.updateConfiguration(ConfigurationManagerImpl.java:691)
        at com.cloud.configuration.ConfigurationManagerImpl.updateConfiguration(ConfigurationManagerImpl.java:869)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:344)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:198)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
        at org.apache.cloudstack.network.contrail.management.EventUtils$EventInterceptor.invoke(EventUtils.java:107)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:175)
        at com.cloud.event.ActionEventInterceptor.invoke(ActionEventInterceptor.java:51)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:175)
        at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:97)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
        at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:215)
        at com.sun.proxy.$Proxy130.updateConfiguration(Unknown Source)
        at org.apache.cloudstack.api.command.admin.config.UpdateCfgCmd.execute(UpdateCfgCmd.java:155)
        at com.cloud.api.ApiDispatcher.dispatch(ApiDispatcher.java:156)
        at com.cloud.api.ApiServer.queueCommand(ApiServer.java:772)
        at com.cloud.api.ApiServer.handleRequest(ApiServer.java:596)
        at com.cloud.api.ApiServlet.processRequestInContext(ApiServlet.java:321)
        at com.cloud.api.ApiServlet$1.run(ApiServlet.java:134)
        at org.apache.cloudstack.managed.context.impl.DefaultManagedContext$1.call(DefaultManagedContext.java:55)
        at org.apache.cloudstack.managed.context.impl.DefaultManagedContext.callWithContext(DefaultManagedContext.java:102)
        at org.apache.cloudstack.managed.context.impl.DefaultManagedContext.runWithContext(DefaultManagedContext.java:52)
        at com.cloud.api.ApiServlet.processRequest(ApiServlet.java:131)
        at com.cloud.api.ApiServlet.doGet(ApiServlet.java:93)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:645)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:750)
        at org.eclipse.jetty.servlet.ServletHolder$NotAsync.service(ServletHolder.java:1443)
        at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:791)
        at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:550)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
        at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:602)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:235)
        at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1624)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)
        at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1435)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)
        at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:501)
        at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1594)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
        at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1350)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
        at org.eclipse.jetty.server.handler.gzip.GzipHandler.handle(GzipHandler.java:766)
        at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:146)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
        at org.eclipse.jetty.server.Server.handle(Server.java:516)
        at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:388)
        at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:633)
        at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:380)
        at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:273)
        at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
        at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
        at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:336)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:313)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:171)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:129)
        at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:375)
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:773)
        at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:905)
        at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.sql.SQLIntegrityConstraintViolationException: Column 'value' cannot be null
        at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:117)
        at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:97)
        at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:122)
        ... 87 more

~~~
![image](https://user-images.githubusercontent.com/12384665/135638946-9a7a12b6-a55b-4ae6-ab4d-ab75e0242673.png)

**Comment by DaanHoogland** [comment]
@vladimirpetrov this wil be solved by the reset values #4230 !?

**Comment by vladimirpetrov** [comment]
Resetting to a default value is different than setting an empty string @DaanHoogland. Currently this empty string is converted to a null value which is rejected by the backend. But I agree, this PR could be a workaround for this particular case with the LDAP settings.

**Comment by rhtyd** [comment]
Please triage @nvazquez - if this is a serious regression requiring fixing?

**Comment by vladimirpetrov** [comment]
Definitely a regression, we don't have it in 4.15.1 @rhtyd @nvazquez

**Comment by sureshanaparti** [comment]
@vladimirpetrov @nvazquez the 'value' column in the domain_details table is altered, and doesn't accept NULL. so, this needs fix.

**Comment by rhtyd** [comment]
Thanks for confirming @sureshanaparti you may want to look at all tables/columns which got altered and preemptively explore/fix potential issues like Vladi has discovered wrt 4.16.0.0.

**Comment by sureshanaparti** [comment]
> Thanks for confirming @sureshanaparti you may want to look at all tables/columns which got altered and preemptively explore/fix potential issues like Vladi has discovered wrt 4.16.0.0.

Fixed @rhtyd, PR here: #5551
